fuzz/fuzz.c

```diff
@@ -0,0 +1,240 @@
+            char* new_buf = realloc(buf, new_smax);
+            if (!new_buf || new_smax <= smax) {
+                free(new_buf ? new_buf : buf);
+                goto error;
+            }
+            smax = new_smax;
+            buf = new_buf;
+        }
+    } while (read_b > 0);
+    buf[s] = '\0';
+    ASS_Track *track = ass_read_memory(ass_library, buf, s, NULL);
+    free(buf);
+    return track;
+error:
+    printf("Input too large!\n");
+    return NULL;
+}
+
+int main(int argc, char *argv[])
+{
+    /* Default failure code of sanitisers is 1, unless
+     * changed via env (A|UB|..)SAN_OPTIONS=exitcode=21
+     */
+    enum {
+        // Everything is fine
+        FUZZ_OK = 0,
+        // default for ASAN and gcc's UBSAN
+        // Clang's UBSAN always uses 0, see https://reviews.llvm.org/D35085
+        //SANITISER_FAIL = 1,
+        // Invalid parameters passed etc
+        FUZZ_BAD_USAGE = 2,
+        // Error before rendering starts
+        FUZZ_INIT_ERR = 0
+    };
+
+    ASS_Track *track = NULL;
+    int retval = FUZZ_OK;
+
+    if (argc != 2) {
+        printf("usage: %s <subtitle file>\n", argc ? argv[0] : "fuzz");
+        return FUZZ_BAD_USAGE;
+    }
+
+    if (!init()) {
+        printf("library init failed!\n");
+        retval = FUZZ_INIT_ERR;
+        goto cleanup;
+    }
+
+    if (strcmp(argv[1], "-"))
+        track = ass_read_file(ass_library, argv[1], NULL);
+    else
+        track = read_track_from_stdin();
+
+    if (!track) {
+        printf("track init failed!\n");
+        retval = FUZZ_INIT_ERR;
+        goto cleanup;
+    }
+
+    consume_track(ass_renderer, track);
+
+cleanup:
+    if (track)        ass_free_track(track);
+    if (ass_renderer) ass_renderer_done(ass_renderer);
+    if (ass_library)  ass_library_done(ass_library);
+
+    return retval;
+}
+#elif ASS_FUZZMODE == FUZZMODE_AFLXX_SHAREDMEM
+__AFL_FUZZ_INIT();
+/*
+ * AFL++ docs recommend to disable optimisation for the main function
+ * and GCC and Clang are the only AFL compilers.
+ */
+#pragma clang optimize off
+#pragma GCC   optimize("O0")
+int main(int argc, char *argv[])
+{
+    // AFLs buffer and length macros should not be used directly
+    ssize_t len;
+    unsigned char *buf;
+
+    if (!init()) {
+        printf("library init failed!\n");
+        return 1;
+    }
+
+    __AFL_INIT();
+    buf = __AFL_FUZZ_TESTCASE_BUF;
+    while (__AFL_LOOP(100000)) {
+        len = __AFL_FUZZ_TESTCASE_LEN;
+        if (len < 50) // Can't be a useful file
+            continue;
+
+        // AFL demo puts cleanup above?
+
+        ASS_Track *track = ass_read_memory(ass_library, (char *)buf, len, NULL);
+        if (!track)
+            continue;
+        consume_track(ass_renderer, track);
+
+        ass_free_track(track);
+        ass_clear_fonts(ass_library);
+        // Renderer caches cannot be cleared (cache_limit just updates the bound values)
+        // Re-calling set_fonts destroys and re-initialises the fontselector at least
```

Comment:
It also clears the font and shaper caches and calls `ass_reconfigure`, which in turn clears the composite, bitmap and outline caches… which, I think, is all of our caches.

And isn’t it unsafe to call `ass_clear_fonts` before clearing the font-related caches? Should that be moved below the `ass_set_fonts` call?

Reply:
Now that you mention it, yes, [FT_New_Memory_Face](https://freetype.org/freetype2/docs/reference/ft2-base_interface.html#ft_new_memory_face)'s notes clearly spell out: “You must not deallocate the memory before calling FT_Done_Face”. Maybe I just got lucky not to run into any problems with it. Switched it around now.